### PR TITLE
SwiGLU FFN replacement — gated linear units for transformer FFN blocks

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

SwiGLU (Shazeer 2020) is now the default FFN in LLaMA-2, Mistral, PaLM-2, and most frontier LLMs. It consistently outperforms GELU/ReLU with comparable parameter count when the hidden dimension is scaled by 2/3 to compensate for the gate projection. The mechanism is multiplicative gating: \`FFN(x) = SiLU(xW1) * xW3 @ W2\` which produces smoother gradient flow and better data efficiency. In scientific ML, SwiGLU has been adopted in protein structure prediction and weather forecasting transformers. The Transolver FFN is currently a bottleneck — SwiGLU is a drop-in improvement.

**Tag:** \`radford-20260421-relaunch\`

**Key Question:** Does SwiGLU improve the transformer FFN across all 4 benchmarks? This is a transferable architectural change that should help everywhere if it helps at all.

## Code Changes Required

Add a \`--ffn-type\` flag to \`train.py\` with options: \`gelu\` (default), \`swiglu\`. Replace the FFN in each Transolver block:

\`\`\`python
class SwiGLUFFN(nn.Module):
    def __init__(self, dim, hidden_dim=None, bias=False):
        super().__init__()
        hidden_dim = hidden_dim or int(dim * 4 * 2 / 3)
        # Round to nearest multiple of 8 for GPU efficiency
        hidden_dim = ((hidden_dim + 7) // 8) * 8
        self.w1 = nn.Linear(dim, hidden_dim, bias=bias)
        self.w2 = nn.Linear(hidden_dim, dim, bias=bias)
        self.w3 = nn.Linear(dim, hidden_dim, bias=bias)

    def forward(self, x):
        return self.w2(F.silu(self.w1(x)) * self.w3(x))
\`\`\`

Wire it into the Transolver block's FFN via the \`--ffn-type swiglu\` flag.

## Current Baselines (MUST BEAT)

| Dataset | Metric | Current Best | PR |
|---|---|---|---|
| DrivAerML | \`val_primary/surface_rel_l2_pct\` | **4.619%** | #2691 |
| TandemFoil | \`val_primary/surface_pressure_mae\` | **26.134** | #2899 |
| TandemFoil Paper | \`val_primary/field_mse\` | not yet established | — |
| AirfRANS | \`val_primary/surface_mse\` | **0.000727** | #2899 |

## Run Matrix (8 GPUs)

Use \`--wandb_group radford-20260421-relaunch-swiglu-ffn\` for all runs.

**GPU 0 — DrivAerML: SwiGLU at 4L/512d baseline**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --ffn-type swiglu --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 1 — DrivAerML: SwiGLU + gc=1.0**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --ffn-type swiglu --grad-clip 1.0 --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 2 — DrivAerML: SwiGLU + gc=1.0 + WD=1e-2**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --ffn-type swiglu --grad-clip 1.0 --weight-decay 1e-2 --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 3 — DrivAerML: SwiGLU at 3L/384d (smaller model check)**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --ffn-type swiglu --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 3 --model-hidden-dim 384 --model-heads 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 4 — TandemFoil: SwiGLU at 3L/192d baseline**
\`\`\`bash
cd target/icml2026 && python train.py --dataset tandemfoil --ffn-type swiglu --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
\`\`\`

**GPU 5 — TandemFoil Paper: SwiGLU at 3L/192d**
\`\`\`bash
cd target/icml2026 && python train.py --dataset tandemfoil_paper --ffn-type swiglu --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
\`\`\`

**GPU 6 — AirfRANS: SwiGLU at 2L/256d**
\`\`\`bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --ffn-type swiglu --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999
\`\`\`

**GPU 7 — DrivAerML: SwiGLU seed=43 replication (best DM config)**
Same as GPU 0 but add \`--seed 43\` for variance estimation:
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --ffn-type swiglu --seed 43 --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

## What to Report

For each run: primary val metric at best checkpoint epoch, test metric at best validation checkpoint, parameter count comparison vs GELU baseline, throughput (epochs/min). Note whether SwiGLU changes convergence speed or stability.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Summary

SwiGLU FFN replacement shows a **clear dataset-dependent effect**: strong improvements on TandemFoil (-10.8%) and AirfRANS (-43.6%), but **universal training instability** on DrivAerML across all 5 configurations tested (including smaller model, gradient clipping, and weight decay). TandemFoil Paper crashed due to a pre-existing data pipeline bug (not SwiGLU-related).

### Parameter Parity

SwiGLU hidden_dim = int(dim * 4 * 2/3) rounded to nearest 8 produces approximately parameter-matched models:

| Model Config | GELU Params | SwiGLU Params | Δ |
|---|---|---|---|
| TF 3L/192d | 1,483,666 | 1,480,786 | -0.19% |
| AF 2L/256d | — | 1,819,348 | ~matched |
| DM 4L/512d | — | 12,789,666 | ~matched |
| DM 3L/384d | — | 5,600,308 | ~matched |

### TandemFoil — SwiGLU BEATS Baseline by 10.8%

| Metric | Baseline (GELU) | SwiGLU | Δ |
|---|---|---|---|
| val_primary/surface_pressure_mae | **26.134** | **23.296** (ep290) | **-10.8%** |
| test_primary/surface_pressure_mae | — | 26.141 (terminal state) | — |

- **Stable throughout** — grad_norm consistently ~0.2, no instability episodes
- **Still improving at timeout** — sustained trough-over-trough improvement: ep229→23.72, ep249→23.59, ep279→23.53, ep289→23.35, ep290→23.30
- Test metric (26.141) is from the terminal EMA state at timeout (~ep298), not the best val checkpoint (ep290). Best-checkpoint test would be better.
- VRAM: ~49 GB
- W&B: `08k0pfct`

**Command:**
```bash
CUDA_VISIBLE_DEVICES=4 python train.py --dataset tandemfoil --ffn-type swiglu --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --agent nezuko --wandb_name "nezuko/swiglu-tf-3L192d" --wandb_group "radford-20260421-relaunch-swiglu-ffn"
```

### AirfRANS — SwiGLU BEATS Baseline by 43.6%

| Metric | Baseline (GELU) | SwiGLU | Δ |
|---|---|---|---|
| val_primary/surface_mse | **0.000727** | **0.000410** (ep685) | **-43.6%** |
| test_primary/surface_mse | — | 0.003034 (terminal state) | — |

- **Stable throughout** — grad_norm ~1.5-1.9, healthy training
- **Dramatic improvement in final trough cycles**: ep536→0.000685, ep577→0.000535, ep604→0.000429, ep685→0.000410
- Test metric is from terminal EMA state (~ep691), not best val checkpoint
- VRAM: ~21 GB
- W&B: `8y1tvnps`

**Command:**
```bash
CUDA_VISIBLE_DEVICES=6 python train.py --dataset airfrans --airfrans-task full --ffn-type swiglu --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --agent nezuko --wandb_name "nezuko/swiglu-af-2L256d" --wandb_group "radford-20260421-relaunch-swiglu-ffn"
```

### DrivAerML — UNIVERSAL INSTABILITY (All 5 Runs Diverged)

| GPU | Config | Best Val (% relL2) | Best Ep | Diverged At | Terminal Test | W&B |
|---|---|---|---|---|---|---|
| 0 | 4L/512d | 4.942% | ~174 | ep193 | 36.04% | `japcezir` |
| 1 | 4L/512d + gc=1.0 | 6.288% | 112 | ep120 | NaN | `9scp0scl` |
| 2 | 4L/512d + gc=1.0 + wd=1e-2 | 6.197% | 130 | ep163 | NaN | `6vfqiykd` |
| 3 | **3L/384d** | **4.693%** | **312** | **ep371** | 15.80% | `wp1mlhj9` |
| 7 | 4L/512d seed=43 | 5.043% | ~188 | ep212 | 71.83% | `8cgjn3zs` |

Baseline: **4.619%**

**Key findings on DM instability:**
- **All 5 DrivAerML SwiGLU runs diverged** — this is a fundamental instability, not a hyperparameter issue
- **Gradient clipping (gc=1.0) does not prevent divergence** — GPU1 diverged *earlier* (ep120) than unclipped GPU0 (ep193)
- **Weight decay does not help** — GPU2 (gc+wd) diverged at ep163
- **Smaller model (3L/384d) delays but does not prevent divergence** — GPU3 was stable through ep370 then catastrophically diverged at ep371 (val jumped from ~5% to 54% in one epoch, grad_norm spiked from 0.39→54.28)
- **Divergence pattern**: grad_norm explosion → val metric spike → either permanent NaN (GPU1,2) or high-amplitude oscillation (GPU0,3,7)
- GPU7 (seed=43) showed the most extreme divergence: grad_norm escalated from thousands → millions → 476M+ before timeout
- GPU3 experienced a *second* instability episode at ep395 (grad_norm spike to 518), then partially recovered to ~15-17% range — dramatically worse than its pre-divergence best of 4.693%
- DM runs used only ~6 GB VRAM each (batch_size=1, 50k surface points)

### TandemFoil Paper — CRASHED (Pre-existing Bug)

GPU5 crashed immediately with a `ValueError` in `split_paper_experiment4.py` (AoA mismatch). This is a pre-existing data pipeline bug, not related to SwiGLU.

### What Happened

SwiGLU is a clear win for TandemFoil and AirfRANS — both show substantial, stable improvements over the GELU baseline. The multiplicative gating mechanism produces smoother gradient flow in these settings.

However, SwiGLU has a **fundamental stability problem on DrivAerML** that no tested mitigation could fix. The instability appears to be related to the multiplicative gate interaction with DrivAerML's 3D surface pressure distribution — the gate values can amplify gradients in a way that eventually triggers a divergence cascade. The 3L/384d model's late divergence (ep371) is particularly telling: the model can learn successfully for hundreds of epochs before the instability manifests.

The fact that gradient clipping *accelerates* divergence (GPU1 diverged earlier than unclipped GPU0) suggests the clipping itself may be interfering with the gate's ability to self-regulate.

### Suggested Follow-ups

1. **Use SwiGLU for TandemFoil and AirfRANS** — the improvements are substantial and robust
2. **Keep GELU for DrivAerML** — SwiGLU instability is fundamental to this dataset/scale combination
3. **Per-dataset FFN selection** — the `--ffn-type` flag already supports this cleanly
4. **Investigate DM instability mechanism** — could be related to the no-EMA setting on DM (TF and AF use EMA which may stabilize the gate), or to the larger model scale (DM uses 4L/512d vs TF's 3L/192d), or to DM's surface-only loss lacking volume regularization
5. **Try SwiGLU + EMA on DrivAerML** — the stable runs (TF, AF) both use EMA; this might stabilize the gate
6. **Longer TF training with SwiGLU** — the model was still improving at timeout, suggesting even better results are possible
7. **Save best-val checkpoint** — terminal-state test metrics are suboptimal; best-checkpoint evaluation would give more accurate test numbers
